operator/pkg/ciliumendpointslice/ceptocesmap.go

```diff
@@ -35,28 +58,180 @@ type CESData struct {
+	c.cepNameToGlobalIdLabels[cepName] = gidLabels
+}
+
+// Update encryption key for node and return all affected CES whose CEPs are on that node.
+func (c *CESToCEPMapping) insertNode(nodeName NodeName, encryptionKey EncryptionKey) []CESKey {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.nodeNameToEncryptionKey[nodeName] = encryptionKey
+	if _, ok := c.nodeNameToCEPNameSet[nodeName]; !ok {
+		c.nodeNameToCEPNameSet[nodeName] = make(map[CEPName]struct{})
+	}
+
+	cesKeys := make([]CESKey, 0, len(c.nodeNameToCEPNameSet[nodeName]))
+	for cepName := range c.nodeNameToCEPNameSet[nodeName] {
+		cesName := c.cepNameToCESName[cepName]
+		cesKeys = append(cesKeys, NewCESKey(cesName.string(), c.cesData[cesName].ns))
+	}
+	return cesKeys
+}
+
+// Insert a CID into the cache and return the CES which need to be reconciled
+func (c *CESToCEPMapping) insertCID(cid CID, gidLabels string) []CESKey {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.globalIdLabelsToCIDSet[gidLabels]; !ok {
+		c.globalIdLabelsToCIDSet[gidLabels] = &SecIDs{
+			ids:  map[CID]struct{}{},
+			ceps: map[CEPName]struct{}{},
+		}
+	}
+	c.globalIdLabelsToCIDSet[gidLabels].ids[cid] = struct{}{}
+
+	cesKeys := make([]CESKey, 0, len(c.globalIdLabelsToCIDSet[gidLabels].ceps))
+	for cepName := range c.globalIdLabelsToCIDSet[gidLabels].ceps {
+		cesName := c.cepNameToCESName[cepName]
+		cesKeys = append(cesKeys, NewCESKey(cesName.string(), c.cesData[cesName].ns))
+	}
+	return cesKeys
 }
 
 // Remove the CEP entry from map
 func (c *CESToCEPMapping) deleteCEP(cepName CEPName) {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
+
+	if node, ok := c.cepNameToNodeName[cepName]; ok {
+		if _, ok := c.nodeNameToCEPNameSet[node]; ok {
+			delete(c.nodeNameToCEPNameSet[node], cepName)
+		}
+	}
+	delete(c.cepNameToNodeName, cepName)
+
+	if labels, ok := c.cepNameToGlobalIdLabels[cepName]; ok {
+		if _, ok := c.globalIdLabelsToCIDSet[labels]; ok {
+			delete(c.globalIdLabelsToCIDSet[labels].ceps, cepName)
+			c.cleanLabelsMap(labels)
+		}
+	}
+	delete(c.cepNameToGlobalIdLabels, cepName)
+
 	delete(c.cesNameToCEPNameSet[c.cepNameToCESName[cepName]], cepName)
 	delete(c.cepNameToCESName, cepName)
 }
 
+// Remove node from cache and return affected CES
+func (c *CESToCEPMapping) deleteNode(nodeName NodeName) []CESKey {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	delete(c.nodeNameToEncryptionKey, nodeName)
+
+	cesKeys := make([]CESKey, 0, len(c.nodeNameToCEPNameSet[nodeName]))
+	for cepName := range c.nodeNameToCEPNameSet[nodeName] {
+		cesName := c.cepNameToCESName[cepName]
+		cesKeys = append(cesKeys, NewCESKey(cesName.string(), c.cesData[cesName].ns))
+	}
+	delete(c.nodeNameToCEPNameSet, nodeName)
+	return cesKeys
+}
+
+// Remove CID from cache and return affected CES
+func (c *CESToCEPMapping) deleteCID(cid CID, gidLabels string) []CESKey {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.globalIdLabelsToCIDSet[gidLabels]; !ok {
+		return []CESKey{}
```

Comment:
Is there a reason to return an empty slice instead of `nil`? Same below at line 211